smokey/src/main/java/frc/robot/common/CorrectableEncoderRevNeoPlusDigitalIoPort.java

```diff
@@ -0,0 +1,80 @@
+// ************************************************************
+// Bishop Blanchet Robotics
+// Home of the Cybears
+// FRC - Charged Up - 2023
+// File: public enum CorrectableEncoderRevNeoPlusDigitalIoPort.java
+// Intent: Forms an interface that will help keep a motor encoder from drifting with use of external sensor.
+// ************************************************************
+
+// ʕ •ᴥ•ʔ ʕ•ᴥ•  ʔ ʕ  •ᴥ•ʔ ʕ •`ᴥ´•ʔ ʕ° •° ʔ ʕ •ᴥ•ʔ ʕ•ᴥ•  ʔ ʕ  •ᴥ•ʔ ʕ •`ᴥ´•ʔ ʕ° •° ʔ 
+
+package frc.robot.common;
+
+import com.revrobotics.RelativeEncoder;
+
+import edu.wpi.first.wpilibj.DigitalInput;
+
+public class CorrectableEncoderRevNeoPlusDigitalIoPort implements ICorrectableEncoder {
+
+    private RelativeEncoder revNeoMotorEncoder = null;
+    private DigitalInput dioStateDevice = null;
+    private double encoderTicksAtStateChangeSetPoint = 0.0;
+    private boolean lastState = false;
+    private boolean motorEncoderPositionReset = false;
+
+    /**
+     * Concrete implementation of a pairing between motor encoder and DIO to coordinate resetting of 
+     * encoder position based on state change of DIO
+     * @param revNeoEncoder - the REV Neo motor encoder
+     * @param stateDevice - the DIO state device
+     * @param encoderTicksAtStateChange - the set point to use for the motors encoder when the state 
+     * @param initalizeEncoderPositionOnDioTriggered - when true and the device is currently triggered the initial position will be set 
+     * changes between the current and previous calls into getCurrentEncoderPosition
+     */
+    public CorrectableEncoderRevNeoPlusDigitalIoPort(
+        RelativeEncoder revNeoEncoder,
+        DigitalInput stateDevice,
+        double encoderTicksAtStateChange,
+        boolean initalizeEncoderPositionOnDioTriggered) {
+
+        revNeoMotorEncoder = revNeoEncoder;
+        dioStateDevice = stateDevice;
+        encoderTicksAtStateChangeSetPoint = encoderTicksAtStateChange;
+
+        lastState = dioStateDevice.get();
+
+        if(initalizeEncoderPositionOnDioTriggered == true && lastState == false) {
```

Comment:
is lastState false when the sensor detects metal, or not?

Reply:
how is this enough information to know you should set the encoder position? I expected to have to see it transition from true->false or vice-versa to know the end of the tub has just passed.

Reply:
So that is in the ctor only ... 

 The ctor is allowing the caller to make the assumption that the sensor being tripped is at one end or the other of the arm travel.  It may be that we find this assumption allows too much leeway (the sensor is not really at the end of travel).  For good or for bad, the old code effectively did this and one intent I had here was to be swapping out to have equivalent/compatible logic with the old code.

Reply:
got it.  I think this is actually a good safe assumption.